src/iceberg/expression/inclusive_metrics_evaluator.cc

```diff
@@ -0,0 +1,544 @@
+      }
+      if (upper_str.starts_with(prefix)) {
+        // both bounds match the prefix, so all rows must match the prefix and therefore
+        // do not satisfy the predicate
+        return ROWS_CANNOT_MATCH;
+      }
+    }
+
+    return ROWS_MIGHT_MATCH;
+  }
+
+ private:
+  bool MayContainNull(int32_t id) {
+    return data_file_.null_value_counts.empty() ||
+           !data_file_.null_value_counts.contains(id) ||
+           data_file_.null_value_counts.at(id) != 0;
+  }
+
+  bool ContainsNullsOnly(int32_t id) {
+    return !data_file_.value_counts.empty() && data_file_.value_counts.contains(id) &&
+           !data_file_.null_value_counts.empty() &&
+           data_file_.null_value_counts.contains(id) &&
+           data_file_.value_counts.at(id) - data_file_.null_value_counts.at(id) == 0;
+  }
+
+  bool ContainsNaNsOnly(int32_t id) {
+    return !data_file_.nan_value_counts.empty() &&
+           data_file_.nan_value_counts.contains(id) && !data_file_.value_counts.empty() &&
+           data_file_.value_counts.at(id) == data_file_.nan_value_counts.at(id);
+  }
+
+  Result<Literal> LowerBound(const std::shared_ptr<BoundTerm>& term) {
+    if (term->kind() == BoundTerm::Kind::kReference) {
+      return ParseLowerBound(*std::dynamic_pointer_cast<BoundReference>(term));
+    } else if (term->kind() == BoundTerm::Kind::kTransform) {
+      return TransformLowerBound(*std::dynamic_pointer_cast<BoundTransform>(term));
+    } else if (term->kind() == BoundTerm::Kind::kExtract) {
+      // TODO(xiao.dong) handle extract lower and upper bounds
+      return NotImplemented("Extract lower bound not implemented.");
+    } else {
+      return NotFound("Lower bound not found.");
+    }
+  }
+
+  Result<Literal> UpperBound(const std::shared_ptr<BoundTerm>& term) {
+    if (term->kind() == BoundTerm::Kind::kReference) {
+      return ParseUpperBound(*std::dynamic_pointer_cast<BoundReference>(term));
+    } else if (term->kind() == BoundTerm::Kind::kTransform) {
+      return TransformUpperBound(*std::dynamic_pointer_cast<BoundTransform>(term));
+    } else if (term->kind() == BoundTerm::Kind::kExtract) {
+      // TODO(xiao.dong) handle extract lower and upper bounds
+      return NotImplemented("Extract upper bound not implemented.");
+    } else {
+      return NotFound("Upper bound not found.");
+    }
+  }
+
+  Result<Literal> ParseLowerBound(const BoundReference& ref) {
+    int id = ref.field().field_id();
+    auto type = ref.type();
+    if (!type->is_primitive()) {
+      return Invalid("Lower bound of non-primitive type is not supported.");
+    }
+    auto primitive_type = std::dynamic_pointer_cast<PrimitiveType>(type);
+    if (!data_file_.lower_bounds.empty() && data_file_.lower_bounds.contains(id)) {
+      ICEBERG_ASSIGN_OR_RAISE(
+          auto lower,
+          Literal::Deserialize(data_file_.lower_bounds.at(id), primitive_type));
+      return lower;
+    }
+
+    return NotFound("Lower bound not found.");
```

Comment:
The `Error` status in the `Result` usually means an error has been found and cannot proceed. We can either change the return type to `Result<std::optional<Literal>>` or return `Literal::Null()` for a missing value (which is a valid case).